### DISCUSSION TITLE
Smoothing Triangle Meshes and WASM and  JS

### DISCUSSION THREAD

**Discussion by hrgdavor** [elalish/manifold#277] [Q&A] [answered]
I was very intrigued by AMF attempt at curved tringles and when read your article https://elalish.blogspot.com/2022/03/smoothing-triangle-meshes.html . I wanted to know more about it in the context of manifold library.

Since I am very attached to doing 3D stuff in JavaScript, I am really want to know more about it, and if maybe I can benefit from it in my JS code. My main focus in on 3D printing and modelling for it by coding parametrizable models.

My mind is exploding with questions by the fact manifold does this stuff internally. 

First thing I am wondering is what king of silly way this could be used. And how these are treated in boolean operations in manifold.

... anyway .. I really want to know more on this topic. Thanks :)

**Comment by elalish** [comment, marked as answer]
I'm so glad you asked! Indeed, I was trying to succeed where AMF failed. Did you try the Scallop example in https://manifoldcad.org/?

The Boolean ops only work on flat triangles, so you need to refine your low-res curved triangles sufficiently before doing operations on them. Thus they are better as inputs to the CAD workflow than for smoothing the end result. I'd eventually like to find a more hybrid, automatic approach, where the Boolean can refine only around intersections as necessary, but that'll take quite a bit of work.

**Reply by hrgdavor** [reply]
I read about scallop, but was unaware it was one of examples as did not take time to go through the examples. I really love the fresh(at least to me) concepts that you are bringing here. Regardless if things are your original ideas or derivatives, it is great that you are bringing it to the spotlight.

Just the concept of having this intermediate format available and manually refining when needed answers many of my needs. If I were to produce refineable objects, I could use less refined for preview and more refined for export. This allows for compromising fidelity for speed when needed. Usually complex models pretty fast start slowing down the code/review/code cycle, so I am thinking ahead, what could I do to re-gain speed and keep iterating on the design.

If the refinement can be stopped after each step and continued, I could pause after X ms show result, and continue working and showing progress until the desired final detail is calculated. 

>a more hybrid, automatic approach, 

that would be huge, but what you have already looks great to me :) .. 

I have not studied your examples in detail, so I may still have some misconceptions. Looking fwd to learning more about manifold lib.

**Reply by elalish** [reply]
Thank you, and yes, I use a similar approach when designing in OpenSCAD, which this is strongly based on. There I tend to set `$fa` high while I work to get a rough preview mode where all primitives have relatively few faces. Then when I'm finished, I set it to a smaller value and get dinner while it computes. Of course this library makes that a lot faster anyway, but with sufficiently complex designs the same situation will inevitably arise. So if you just set your `refine(n)` value and `setMinCircularAngle` at the top of your script, that gives you the same concept.
